README.md
# ripple-keypairs-rust
[![Crates.io: ripple-keypairs](https://img.shields.io/crates/v/ripple-keypairs.svg)](https://crates.io/crates/ripple-keypairs)
[![Documentation](https://docs.rs/ripple-keypairs/badge.svg)](https://docs.rs/ripple-keypairs)
[![CI Status (Github Actions)](https://github.com/otov4its/ripple-keypairs-rust/workflows/CI/badge.svg?master)](https://github.com/otov4its/ripple-keypairs-rust/actions)
[![CD Status (Github Actions)](https://github.com/otov4its/ripple-keypairs-rust/workflows/CD/badge.svg?master)](https://github.com/otov4its/ripple-keypairs-rust/actions)

_\* ripple-keypairs-rust is a **Rust port** of the original JavaScript/TypeScript [ripple-keypairs](https://github.com/ripple/ripple-keypairs)_

An implementation of XRP Ledger keypairs & wallet generation which supports rfc6979 and eddsa  eterministic signatures.

## Documentation

[Documentation](https://docs.rs/ripple-keypairs/)

## Release history

See [HISTORY.md](HISTORY.md)

## Tests

```bash
$ cargo test
```

## License

Licensed under ISC license [LICENSE](LICENSE)


HISTORY.md
# Release History

## Unreleased

## 0.1.0 (2021-01-25)

### Initial release


tests/api.rs
use std::str::FromStr;

use ripple_keypairs::{
    error::{DecodeError, InvalidSignature},
    Algorithm::{Ed25519, Secp256k1},
    Entropy::{Array, Random},
    EntropyArray, HexBytes, Seed,
};

use fixtures::*;

mod fixtures {
    use super::*;

    pub const ENTROPY: EntropyArray = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16];

    pub struct TestData {
        pub entropy: EntropyArray,
        pub seed: &'static str,
        pub private_key: &'static str,
        pub public_key: &'static str,
        pub address: &'static str,
        pub message: &'static str,
        pub signature: &'static str,
    }

    pub static TEST_SECP256K1: TestData = TestData {
        entropy: ENTROPY,
        seed: "sp5fghtJtpUorTwvof1NpDXAzNwf5",
        private_key: "00D78B9735C3F26501C7337B8A5727FD53A6EFDBC6AA55984F098488561F985E23",
        public_key: "030D58EB48B4420B1F7B9DF55087E0E29FEF0E8468F9A6825B01CA2C361042D435",
        address: "rU6K7V3Po4snVhBBaU29sesqs2qTQJWDw1",
        message: "test message",
        signature: "30440220583A91C95E54E6A651C47BEC22744E0B101E2C4060E7B08F6341657DAD9BC3EE02207D1489C7395DB0188D3A56A977ECBA54B36FA9371B40319655B1B4429E33EF2D"
    };

    pub static TEST_ED25519: TestData = TestData {
        entropy: ENTROPY,
        seed: "sEdSKaCy2JT7JaM7v95H9SxkhP9wS2r",
        private_key: "EDB4C4E046826BD26190D09715FC31F4E6A728204EADD112905B08B14B7F15C4F3",
        public_key: "ED01FA53FA5A7E77798F882ECE20B1ABC00BB358A9E55A202D0D0676BD0CE37A63",
        address: "rLUEXYuLiQptky37CqLcm9USQpPiz5rkpD",
        message: "test message",
        signature: "CB199E1BFD4E3DAA105E4832EEDFA36413E1F44205E4EFB9E27E826044C21E3E2E848BBC8195E8959BADF887599B7310AD1B7047EF11B682E0D068F73749750E"
    };

    pub fn secp256k1_test_seed() -> Seed {
        Seed::from_str(TEST_SECP256K1.seed).unwrap()
    }

    pub fn ed25519_test_seed() -> Seed {
        Seed::from_str(TEST_ED25519.seed).unwrap()
    }
}

mod secp256k1 {
    use super::*;

    #[test]
    fn new_seed() {
        let seed = Seed::new(Array(TEST_SECP256K1.entropy), &Secp256k1);

        assert_eq!(TEST_SECP256K1.seed, seed.to_string());
    }

    #[test]
    fn random_seed_starts_with_s() {
        let seed = Seed::random();

        assert!(seed.to_string().starts_with("s") && !seed.to_string().starts_with("sEd"));
    }

    #[test]
    fn random_seed() {
        assert_ne!(Seed::random().to_string(), Seed::random().to_string());
    }

    #[test]
    fn parse_random_seed() {
        let random_seed = Seed::random();
        let parsed_seed: Seed = random_seed.to_string().parse().unwrap();

        assert_eq!(parsed_seed, random_seed);
    }

    #[test]
    fn parse_seed() {
        let seed: Seed = TEST_SECP256K1.seed.parse().unwrap();

        assert_eq!(seed.as_kind(), &Secp256k1);
        assert_eq!(seed.as_entropy(), &TEST_SECP256K1.entropy);
    }

    #[test]
    fn bad_seed() {
        assert_eq!(
            "sXXXghtJtpUorTwvof1NpDXAzNwf5".parse::<Seed>().unwrap_err(),
            DecodeError
        );
    }

    #[test]
    fn derive_keypair() {
        let seed = secp256k1_test_seed();
        let (private, public) = seed.derive_keypair().unwrap();

        assert_eq!(
            (private.to_string().as_str(), public.to_string().as_str()),
            (TEST_SECP256K1.private_key, TEST_SECP256K1.public_key)
        );
    }

    #[test]
    fn sign() {
        let seed = secp256k1_test_seed();
        let (private, _) = seed.derive_keypair().unwrap();

        assert_eq!(
            private.sign(&TEST_SECP256K1.message).to_string(),
            TEST_SECP256K1.signature
        )
    }

    #[test]
    fn verify() {
        let seed = secp256k1_test_seed();
        let (_, public) = seed.derive_keypair().unwrap();

        let sig = HexBytes::from_hex_unchecked(TEST_SECP256K1.signature);

        assert_eq!(public.verify(&TEST_SECP256K1.message, &sig), Ok(()));
    }

    #[test]
    fn verify_bad_signature() {
        let seed = secp256k1_test_seed();
        let (_, public) = seed.derive_keypair().unwrap();

        assert_eq!(
            public.verify(&TEST_SECP256K1.message, &"bad signature"),
            Err(InvalidSignature)
        );
    }

    #[test]
    fn derive_address() {
        let (_, public) = secp256k1_test_seed().derive_keypair().unwrap();

        assert_eq!(public.derive_address(), TEST_SECP256K1.address);
    }

    #[test]
    fn random_address() {
        let random_seed = Seed::new(Random, &Secp256k1);
        let (_, public) = random_seed.derive_keypair().unwrap();
        let address = public.derive_address();

        assert_eq!("r", &address[..1]);
    }
}

mod ed25519 {
    use super::*;

    #[test]
    fn new_seed() {
        let seed = Seed::new(Array(TEST_ED25519.entropy), &Ed25519);

        assert_eq!(TEST_ED25519.seed, seed.to_string());
    }

    #[test]
    fn random_seed_starts_with_sed() {
        let seed = Seed::new(Random, &Ed25519);

        assert!(seed.to_string().starts_with("sEd"));
    }

    #[test]
    fn random_seed() {
        assert_ne!(
            Seed::new(Random, &Ed25519).to_string(),
            Seed::new(Random, &Ed25519).to_string()
        );
    }

    #[test]
    fn parse_random_seed() {
        let random_seed = Seed::new(Random, &Ed25519);
        let parsed_seed: Seed = random_seed.to_string().parse().unwrap();

        assert_eq!(parsed_seed.as_kind(), random_seed.as_kind());
        assert_eq!(parsed_seed.as_entropy(), random_seed.as_entropy());
    }

    #[test]
    fn parse_seed() {
        let seed: Seed = TEST_ED25519.seed.parse().unwrap();

        assert_eq!(seed.as_kind(), &Ed25519);
        assert_eq!(seed.as_entropy(), &TEST_ED25519.entropy);
    }

    #[test]
    fn bad_seed() {
        assert_eq!(
            "sEdXXXCy2JT7JaM7v95H9SxkhP9wS2r"
                .parse::<Seed>()
                .unwrap_err(),
            DecodeError
        );
    }

    #[test]
    fn derive_keypair() {
        let seed = ed25519_test_seed();
        let (private, public) = seed.derive_keypair().unwrap();

        assert_eq!(
            (private.to_string().as_str(), public.to_string().as_str()),
            (TEST_ED25519.private_key, TEST_ED25519.public_key)
        );
    }

    #[test]
    fn sign() {
        let seed = ed25519_test_seed();
        let (private, _) = seed.derive_keypair().unwrap();

        assert_eq!(
            private.sign(&TEST_ED25519.message).to_string(),
            TEST_ED25519.signature
        )
    }

    #[test]
    fn verify() {
        let seed = ed25519_test_seed();
        let (_, public) = seed.derive_keypair().unwrap();

        let sig = HexBytes::from_hex_unchecked(TEST_ED25519.signature);

        assert_eq!(public.verify(&TEST_ED25519.message, &sig), Ok(()));
    }

    #[test]
    fn verify_bad_signature() {
        let seed = ed25519_test_seed();
        let (_, public) = seed.derive_keypair().unwrap();

        assert_eq!(
            public.verify(&TEST_ED25519.message, &"bad signature"),
            Err(InvalidSignature)
        );
    }

    #[test]
    fn derive_address() {
        let (_, public) = ed25519_test_seed().derive_keypair().unwrap();

        assert_eq!(public.derive_address(), TEST_ED25519.address);
    }

    #[test]
    fn random_address() {
        let random_seed = Seed::new(Random, &Ed25519);
        let (_, public) = random_seed.derive_keypair().unwrap();
        let address = public.derive_address();

        assert_eq!("r", &address[..1]);
    }
}


src/error.rs
//! Errors

use std::{error, fmt, result};

use base_x;

pub use Error::*;

/// Result with error type
pub type Result<T> = result::Result<T, Error>;

/// Error enum type
#[allow(missing_docs)]
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum Error {
    ExpectedMinEntorpyLenght(usize),
    DecodeError,
    InvalidSignature,
    DeriveKeyPairError,
}

impl error::Error for Error {
    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
        None
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ExpectedMinEntorpyLenght(n) => write!(f, "{} {}", "entropy lenght must be >=", n),
            DecodeError => f.write_str("decode error"),
            InvalidSignature => f.write_str("invalid signature"),
            DeriveKeyPairError => f.write_str("derive keypair error"),
        }
    }
}

macro_rules! impl_from_error {
    ($t:ty => $m:ident) => {
        #[doc(hidden)]
        impl From<$t> for Error {
            fn from(_: $t) -> Self {
                $m
            }
        }
    };
}

impl_from_error!(base_x::DecodeError => DecodeError);
impl_from_error!(secp256k1::Error => InvalidSignature);


src/hexbytes.rs
use std::{fmt, str::FromStr};

use hex::{decode, encode_upper};

use crate::error;

/// The values of this type can be treated as bytes or as hex encoded strings
///
/// Mostly used for crate's internal puprposes,
/// but can also be used as a hex encode/decode utility.
///
/// # Examples
///
/// ```
/// # use std::error::Error;
/// # fn main() -> Result<(), Box<dyn Error>> {
/// #
/// use ripple_keypairs::HexBytes;
///
/// let hexbytes = HexBytes::from_hex_unchecked("00");
///
/// assert_eq!(hexbytes.as_bytes(), &[0]);
/// assert_eq!(hexbytes.as_hex(), "00");
/// assert_eq!(hexbytes, HexBytes::from_bytes(&[0]));
/// #
/// # Ok(())
/// # }
/// ```
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct HexBytes {
    bytes: Vec<u8>,
    hex_string: String,
}

impl HexBytes {
    /// Construct from raw bytes
    pub fn from_bytes(bytes: &impl AsRef<[u8]>) -> Self {
        let bytes = bytes.as_ref();
        Self {
            bytes: bytes.to_owned(),
            hex_string: encode_upper(bytes),
        }
    }

    /// Construct from hex string
    ///
    /// # Panics
    ///
    /// Need a hex encoded string, panic otherwise.
    pub fn from_hex_unchecked(hex_string: &str) -> Self {
        Self::from_hex(hex_string).expect("need a hex encoded string")
    }

    /// Construct from hex string
    ///
    /// # Errors
    ///
    /// [`error::DecodeError`] if a string is not hex encoded.
    pub fn from_hex(hex_string: &str) -> error::Result<Self> {
        let bytes = decode(hex_string).map_err(|_| error::DecodeError)?;

        Ok(Self {
            bytes,
            hex_string: hex_string.to_owned(),
        })
    }

    /// Treat as bytes
    pub fn as_bytes(&self) -> &[u8] {
        &self.bytes
    }

    /// Treat as a hex encoded string
    pub fn as_hex(&self) -> &str {
        &self.hex_string
    }
}

impl fmt::Display for HexBytes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.as_hex())
    }
}

impl From<&[u8]> for HexBytes {
    fn from(bytes: &[u8]) -> Self {
        Self::from_bytes(&bytes)
    }
}

impl FromStr for HexBytes {
    type Err = error::Error;

    fn from_str(s: &str) -> error::Result<Self> {
        Self::from_hex(s)
    }
}

impl AsRef<[u8]> for HexBytes {
    fn as_ref(&self) -> &[u8] {
        self.as_bytes()
    }
}

impl AsRef<str> for HexBytes {
    fn as_ref(&self) -> &str {
        self.as_hex()
    }
}

impl From<HexBytes> for Vec<u8> {
    fn from(hex_bytes: HexBytes) -> Self {
        hex_bytes.bytes
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hex_to_bytes() {
        assert_eq!(HexBytes::from_hex("000000").unwrap().as_bytes(), &[0, 0, 0]);
    }

    #[test]
    fn bytes_to_hex() {
        assert_eq!(
            HexBytes::from_bytes(&[222, 173, 190, 239]).to_string(),
            "DEADBEEF"
        );
    }

    #[test]
    fn bytes_into_hex_bytes() {
        let hex: HexBytes = [222, 173, 190, 239].as_ref().into();

        assert_eq!(hex.as_bytes(), &[222, 173, 190, 239]);
        assert_eq!(hex.as_hex(), "DEADBEEF");
    }

    #[test]
    fn parse() {
        let hex = "00".parse::<HexBytes>().unwrap();

        assert_eq!(hex.as_hex(), "00");
        assert_eq!(hex.as_bytes(), &[0]);
    }

    #[test]
    fn empty_bytes() {
        let hex = HexBytes::from_bytes(&[]);

        assert_eq!(hex.as_bytes(), &[]);
        assert_eq!(hex.as_hex(), "");
    }

    #[test]
    fn empty_hex_string() {
        let hex = HexBytes::from_hex_unchecked("");

        assert_eq!(hex.as_bytes(), &[]);
        assert_eq!(hex.as_hex(), "");
    }
}


src/algorithm/mod.rs
use std::fmt;

use crate::{error::Result, EntropyArray, HexBytes, KeyPairResult};

pub(super) mod ed25519;
pub(super) mod secp256k1;

pub(super) trait Key {
    fn key_lenght(&self) -> usize;

    fn prefix(&self) -> &[u8];

    fn as_bytes<'a>(&self, bytes: &'a [u8]) -> &'a [u8] {
        &bytes[..self.key_lenght()]
    }

    fn encode_to_hex(&self, bytes: &[u8]) -> String {
        HexBytes::from_bytes(&[self.prefix(), bytes].concat()).to_string()
    }
}

pub(super) trait Sign: Key + fmt::Debug {
    fn sign(&self, message: &[u8], private_key: &[u8]) -> HexBytes;
}

pub(super) trait Verify: Key + fmt::Debug {
    fn verify(&self, message: &[u8], signature: &[u8], public_key: &[u8]) -> Result<()>;
}

pub(super) trait Seed: fmt::Debug {
    fn derive_keypair(&self, entropy: &EntropyArray) -> KeyPairResult;

    fn encode(&self, entropy: &EntropyArray) -> String;
}


src/algorithm/ed25519.rs
use ring::signature::{self, KeyPair};

use ripple_address_codec as codec;

use crate::{
    error::{DeriveKeyPairError, InvalidSignature, Result},
    utils, Ed25519, EntropyArray, HexBytes, KeyPairResult, PrivateKey, PublicKey,
};

use super::{Key, Seed, Sign, Verify};

#[derive(Debug)]
pub(crate) struct PrivateKeyEd25519;

impl Sign for PrivateKeyEd25519 {
    fn sign(&self, message: &[u8], private_key: &[u8]) -> HexBytes {
        let key_pair = signature::Ed25519KeyPair::from_seed_unchecked(private_key).unwrap();

        HexBytes::from_bytes(&key_pair.sign(message))
    }
}

impl Key for PrivateKeyEd25519 {
    fn key_lenght(&self) -> usize {
        Self::LENGHT
    }

    fn prefix(&self) -> &[u8] {
        Self::PREFIX
    }
}

impl PrivateKeyEd25519 {
    const LENGHT: usize = 32;
    const PREFIX: &'static [u8] = &[0xED];
}

#[derive(Debug)]
pub(crate) struct PublicKeyEd25519;

impl Verify for PublicKeyEd25519 {
    fn verify(&self, message: &[u8], signature: &[u8], public_key: &[u8]) -> Result<()> {
        let public_key = signature::UnparsedPublicKey::new(&signature::ED25519, public_key);

        public_key
            .verify(message, signature)
            .map_err(|_| InvalidSignature)
    }
}

impl Key for PublicKeyEd25519 {
    fn key_lenght(&self) -> usize {
        Self::LENGHT
    }

    fn prefix(&self) -> &[u8] {
        Self::PREFIX
    }
}

impl PublicKeyEd25519 {
    const LENGHT: usize = 32;
    const PREFIX: &'static [u8] = &[0xED];
}

#[derive(Debug)]
pub(crate) struct SeedEd25519;

impl Seed for SeedEd25519 {
    fn derive_keypair(&self, entropy: &EntropyArray) -> KeyPairResult {
        let raw_priv = utils::sha512_digest_32(entropy);

        let key_pair = signature::Ed25519KeyPair::from_seed_unchecked(&raw_priv)
            .map_err(|_| DeriveKeyPairError)?;

        let raw_pub = key_pair.public_key().as_ref().to_vec();

        let kind = &Ed25519;

        Ok((
            PrivateKey {
                bytes: raw_priv,
                kind,
            },
            PublicKey {
                bytes: raw_pub,
                kind,
            },
        ))
    }

    fn encode(&self, entropy: &EntropyArray) -> String {
        codec::encode_seed(entropy, &Ed25519)
    }
}


src/lib.rs
//! Cryptographic key pairs for the XRP Ledger
//!
//! An implementation of XRP Ledger keypairs & wallet generation
//! which supports rfc6979 and eddsa deterministic signatures.
//!
//! # Examples
//!
//! ## Generate a random XRP Ledger address
//!
//! ```
//! # use std::error::Error;
//! # fn main() -> Result<(), Box<dyn Error>> {
//! #
//! use ripple_keypairs::Seed;
//!
//! let seed = Seed::random();
//! let (_, public_key) = seed.derive_keypair()?;
//! let address = public_key.derive_address();
//!
//! assert!(address.starts_with("r"));
//! #
//! # Ok(())
//! # }
//! ```
//!
//! ## Encode a seed in Base58 XRP Legder format
//!
//! ```
//! use ripple_keypairs::{Seed, Entropy, Algorithm};
//!
//! // In the real world you **must** generate random entropy
//! let entropy = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16];
//! let seed = Seed::new(Entropy::Array(entropy), &Algorithm::Secp256k1);
//!
//! assert_eq!(seed.to_string(), "sp5fghtJtpUorTwvof1NpDXAzNwf5");
//! ```
//!
//! ## Parse a string into a seed
//!
//! ```
//! # use std::error::Error;
//! # fn main() -> Result<(), Box<dyn Error>> {
//! #
//! use std::str::FromStr;
//! use ripple_keypairs::{Seed, error};
//!
//! let seed = Seed::from_str("sp5fghtJtpUorTwvof1NpDXAzNwf5")?;
//!
//! assert_eq!(seed, "sp5fghtJtpUorTwvof1NpDXAzNwf5".parse()?);
//! assert_eq!(Err(error::DecodeError), "bad seed".parse::<Seed>());
//! #
//! # Ok(())
//! # }
//! ```

#![deny(
    warnings,
    clippy::all,
    missing_debug_implementations,
    missing_copy_implementations,
    missing_docs,
    missing_crate_level_docs,
    missing_doc_code_examples,
    non_ascii_idents,
    unreachable_pub
)]
#![doc(test(attr(deny(warnings))))]
#![doc(html_root_url = "https://docs.rs/ripple-keypairs/0.1.0")]

use std::{convert::TryInto, fmt, str::FromStr};

use ring::rand::{SecureRandom, SystemRandom};

mod utils;

mod hexbytes;
pub use hexbytes::HexBytes;

mod algorithm;
use algorithm as alg;

pub mod error;
type KeyPairResult = error::Result<(PrivateKey, PublicKey)>;

pub use codec::{Algorithm, Entropy as EntropyArray};
use ripple_address_codec as codec;

use Algorithm::*;
use Entropy::*;

/// Entropy which is used to generate seed
#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
pub enum Entropy {
    /// Random entropy
    Random,
    /// An array of bytes
    Array(EntropyArray),
}

/// A seed that can be used to generate keypairs
///
/// # Examples
///
/// ## Generate a new seed
///
/// ```
/// use ripple_keypairs::{Seed, Entropy, Algorithm};
///
/// let seed = Seed::new(Entropy::Random, &Algorithm::Secp256k1);
///
/// assert!(seed.to_string().starts_with("s"));
/// ```
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Seed {
    entropy: EntropyArray,
    kind: &'static Algorithm,
}

impl Seed {
    /// Generate a new seed
    ///
    /// # Examples
    ///
    /// ```
    /// use ripple_keypairs::{Seed, Entropy, Algorithm};
    ///
    /// let seed_secp256k1 = Seed::new(Entropy::Random, &Algorithm::Secp256k1);
    ///
    /// assert!(seed_secp256k1.to_string().starts_with("s"));
    /// assert_eq!(seed_secp256k1.as_kind(), &Algorithm::Secp256k1);
    ///
    /// let seed_ed25519 = Seed::new(Entropy::Random, &Algorithm::Ed25519);
    ///
    /// assert!(seed_ed25519.to_string().starts_with("s"));
    /// assert_eq!(seed_ed25519.as_kind(), &Algorithm::Ed25519);
    /// ```
    ///
    /// # Panics
    ///
    /// Panics only if something goes wrong with the random generator
    /// when using the [`Entropy::Random`] parameter.
    pub fn new(entropy: Entropy, kind: &'static Algorithm) -> Self {
        let entropy = match entropy {
            Array(entropy) => entropy,

            Random => {
                let mut entropy: EntropyArray = [0; 16];

                SystemRandom::new()
                    .fill(&mut entropy)
                    .expect("unspecified random geterator error");

                entropy
            }
        };

        Self { entropy, kind }
    }

    /// Generate a random seed
    ///
    /// The algorithm defaults to Secp256k1.
    ///
    /// # Examples
    ///
    /// ```
    /// use ripple_keypairs::{Seed, Algorithm};
    ///
    /// let seed = Seed::random();
    ///
    /// assert_eq!(seed.as_kind(), &Algorithm::Secp256k1);
    /// assert_ne!(Seed::random(), Seed::random());
    /// ```
    pub fn random() -> Self {
        Self::new(Random, &Secp256k1)
    }

    /// Derive a public and private key from a seed
    ///
    /// # Examples
    ///
    /// ```
    /// # use std::error::Error;
    /// # fn main() -> Result<(), Box<dyn Error>> {
    /// #
    /// use ripple_keypairs::Seed;
    ///
    /// let seed = Seed::random();
    /// let (private_key, public_key) = seed.derive_keypair()?;
    /// let msg = "Test message";
    ///
    /// assert_eq!(public_key.verify(&msg, &private_key.sign(&msg)), Ok(()));
    /// #
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// # Errors
    ///
    /// May return [`error::DeriveKeyPairError`] if the derived keypair
    /// did not generate a verifiable signature
    pub fn derive_keypair(&self) -> KeyPairResult {
        let keypair = self.method().derive_keypair(self.as_entropy())?;

        /* additional safety check */
        {
            let test_message =
                utils::sha512_digest_32("This test message should verify".as_bytes());

            let (private_key, public_key) = &keypair;

            public_key
                .verify(&test_message, &private_key.sign(&test_message))
                .map_err(|_| error::DeriveKeyPairError)?;
        }

        Ok(keypair)
    }

    /// Seed as [`EntropyArray`]
    ///
    /// # Examples
    /// ```
    /// use ripple_keypairs::{Seed, Entropy, Algorithm, EntropyArray};
    ///
    /// let seed = Seed::new(Entropy::Array([0; 16]), &Algorithm::Secp256k1);
    ///
    /// assert_eq!(seed.as_entropy(), &[0; 16]);
    /// assert_eq!(seed.as_entropy(), <Seed as AsRef<EntropyArray>>::as_ref(&seed));
    /// ```
    ///
    /// # Traits
    ///
    /// This method is used in [`AsRef`] trait.
    pub fn as_entropy(&self) -> &EntropyArray {
        &self.entropy
    }

    /// Seed as [`Algorithm`]
    ///
    /// # Examples
    /// ```
    /// use ripple_keypairs::{Seed, Entropy, Algorithm};
    ///
    /// let seed = Seed::new(Entropy::Random, &Algorithm::Ed25519);
    ///
    /// assert_eq!(seed.as_kind(), &Algorithm::Ed25519);
    /// assert_eq!(seed.as_kind(), <Seed as AsRef<Algorithm>>::as_ref(&seed));
    /// ```
    ///
    /// # Traits
    ///
    /// This method is used in [`AsRef`] trait.
    pub fn as_kind(&self) -> &Algorithm {
        self.kind
    }

    fn method(&self) -> &'static dyn alg::Seed {
        match self.kind {
            Secp256k1 => &alg::secp256k1::SeedEcDsaSecP256K1,
            Ed25519 => &alg::ed25519::SeedEd25519,
        }
    }
}

impl fmt::Display for Seed {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.method().encode(&self.entropy))
    }
}

impl FromStr for Seed {
    type Err = error::Error;

    fn from_str(s: &str) -> error::Result<Self> {
        let (entropy, kind) = codec::decode_seed(s).map_err(|_| error::DecodeError)?;

        Ok(Self::new(Array(entropy), kind))
    }
}

impl AsRef<EntropyArray> for Seed {
    fn as_ref(&self) -> &EntropyArray {
        self.as_entropy()
    }
}

impl AsRef<Algorithm> for Seed {
    fn as_ref(&self) -> &Algorithm {
        self.as_kind()
    }
}

/// Signatures can be treated as bytes or as hex encoded strings.
pub trait Signature: AsRef<[u8]> + AsRef<str> + ToString + Into<Vec<u8>> {}

impl Signature for HexBytes {}

/// A private key that can be used to sign messages
///
/// # Examples
///
/// ```
/// # use std::error::Error;
/// # fn main() -> Result<(), Box<dyn Error>> {
/// #
/// use ripple_keypairs::Seed;
///
/// let seed = "sp5fghtJtpUorTwvof1NpDXAzNwf5".parse::<Seed>()?;
///
/// let (private_key, _) = seed.derive_keypair()?;
///
/// let signature = private_key.sign(&"test message");
///
/// assert_eq!(signature.to_string(), "30440220583A91C95E54E6A651C47BEC22744E0B101E2C4060E7B08F6341657DAD9BC3EE02207D1489C7395DB0188D3A56A977ECBA54B36FA9371B40319655B1B4429E33EF2D");
/// assert_eq!(signature.into(), vec![48, 68, 2, 32, 88, 58, 145, 201, 94, 84, 230, 166, 81, 196, 123, 236, 34, 116, 78, 11, 16, 30, 44, 64, 96, 231, 176, 143, 99, 65, 101, 125, 173, 155, 195, 238, 2, 32, 125, 20, 137, 199, 57, 93, 176, 24, 141, 58, 86, 169, 119, 236, 186, 84, 179, 111, 169, 55, 27, 64, 49, 150, 85, 177, 180, 66, 158, 51, 239, 45]);
///
/// assert_eq!(private_key.to_string(), "00D78B9735C3F26501C7337B8A5727FD53A6EFDBC6AA55984F098488561F985E23");
/// #
/// # Ok(())
/// # }
/// ```
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct PrivateKey {
    bytes: Vec<u8>,
    kind: &'static Algorithm,
}

impl PrivateKey {
    /// Sign message
    ///
    /// Returns the [`Signature`] which that can be treated
    /// as bytes or as a hex encoded string.
    pub fn sign(&self, message: &impl AsRef<[u8]>) -> impl Signature {
        self.method()
            .sign(message.as_ref(), &self.method().as_bytes(&self.bytes))
    }

    fn method(&self) -> &'static dyn alg::Sign {
        match self.kind {
            Secp256k1 => &alg::secp256k1::PrivateKeyEcDsaSecP256K1,
            Ed25519 => &alg::ed25519::PrivateKeyEd25519,
        }
    }
}

impl fmt::Display for PrivateKey {
    /// Display as a hex encoded string
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            self.method()
                .encode_to_hex(&self.method().as_bytes(&self.bytes))
        )
    }
}

/// A public key that can be used to derive an XRP Ledger classic address and verify signatures
///
/// # Examples
///
/// ```
/// # use std::error::Error;
/// # fn main() -> Result<(), Box<dyn Error>> {
/// #
/// use ripple_keypairs::{Seed, error};
///
/// let seed = "sp5fghtJtpUorTwvof1NpDXAzNwf5".parse::<Seed>()?;
///
/// let (private_key, public_key) = seed.derive_keypair()?;
///
/// assert_eq!(public_key.derive_address(), "rU6K7V3Po4snVhBBaU29sesqs2qTQJWDw1");
///
/// assert_eq!(public_key.to_string(), "030D58EB48B4420B1F7B9DF55087E0E29FEF0E8468F9A6825B01CA2C361042D435");
///
/// let msg = "Test message";
///
/// assert_eq!(public_key.verify(&msg, &private_key.sign(&msg)), Ok(()));
/// assert_eq!(public_key.verify(&msg, &"bad signature"), Err(error::InvalidSignature));
/// #
/// # Ok(())
/// # }
/// ```
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct PublicKey {
    bytes: Vec<u8>,
    kind: &'static Algorithm,
}

impl PublicKey {
    /// Verify a signature
    ///
    /// # Errors
    ///
    /// Returns [`error::InvalidSignature`] if the signature is invalid.
    pub fn verify(
        &self,
        message: &impl AsRef<[u8]>,
        signature: &impl AsRef<[u8]>,
    ) -> error::Result<()> {
        self.method().verify(
            message.as_ref(),
            signature.as_ref(),
            &self.method().as_bytes(&self.bytes),
        )
    }

    /// Derive an XRP Ledger classic address
    pub fn derive_address(&self) -> String {
        let hex = HexBytes::from_hex_unchecked(&self.to_string());
        let hash: [u8; 20] = utils::hash160(&hex.as_bytes())[..20].try_into().unwrap();

        codec::encode_account_id(&hash)
    }

    fn method(&self) -> &'static dyn alg::Verify {
        match self.kind {
            Secp256k1 => &alg::secp256k1::PublicKeyEcDsaSecP256K1,
            Ed25519 => &alg::ed25519::PublicKeyEd25519,
        }
    }
}

impl fmt::Display for PublicKey {
    /// Display as a hex encoded string
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            self.method()
                .encode_to_hex(&self.method().as_bytes(&self.bytes))
        )
    }
}
